mantidproject/mantid
# Add title-matcher field in the refl gui lookup table

**Description of work.**

This PR adds the new title-matcher field to the experiment settings lookup table. This involes quite a lot of under-the-hood work that needs to be done in one hit, so this addresses all of these issues:
- Add the new field to the view (#33513)
- Add validation for the new field (#33563)
- Add the new field to the encoder/decoder and support old file versions (#33514)

Note that quite a bit of renaming has been done - in the Encoder/Decoder this is to fix cppcheck warnings; in the lookup table it is to generalise names to refer to search criteria more generally, rather than theta-search.

Fixes #33513 and fixes #33563 and fixes #33514

**To test:**

*Basic tests*

- Open the ISIS Reflectometry interface
- Add two groups called `Group1` and `Group2` both with 2 rows:
  - one containing run `13460` and angle `0.5`
  - the other with run `13462` and angle `2.3`
- Click process. They should turn green and the `dq/q` should be around `0.048` and `0.034` respectively.
- Go to the Experiment Settings tab. In the lookup table try entering various values in the lookup rows. Re-process the runs each time. The dq/q should be updated to match the one set in the lookup row when the angles and titles both match, but should use the default dq/q when the criteria do not match.
  - Try entering dq/q only, i.e. angle and title are empty. This is a "wildcard" row and should match all runs
  - Try entering an angle e.g. `0.5`. This should only be applied to runs with angle `0.5`. It should override any wildcard row.
  - Try entering a title as well as an angle. This should only be applied to runs whose group name and angle both match.

*Variations*

- Try with angles within `0.01` of the run value - this is a tolerance used in matching so values within this difference should still match.
- Try with a regex instead of a simple string in the title  matcher
- Click Restore Defaults and check that the lookup table is reverted to its default state (note that some instruments contain values in the table by default but others don't)
- Change the instrument and check the table is restored to its defaults.

*Error conditions*

- Try entering an invalid regex e.g. `[` - the field should be highlighted red
- Trying to process runs when any error is displayed in the lookup table should be disabled (you'll get a popup warning)
- Try entering the same search criteria in two rows - this is an error
- Try entering a regex on a row without an angle - this is an error
- Try entering the same angle twice but with different regexes (or one set and one empty) - this should be ok and should work.

*Save and load*

- Enter some rows on the lookup table.
- Go to the Batch menu and click Save. Save the batch to a file.
- Close the batch tab (click x on the tab on the left) or open a new, empty batch tab. Go to Batch->Load and reload the file into a new batch. Check that the values are re-loaded.
- Try saving a batch from a previous version of mantid and load it into this new version. The previous table values should all be loaded successfully and the title-matcher field will be empty for all rows. 





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
